src/dbnode/storage/index/wide_query_results.go

```diff
@@ -0,0 +1,274 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+// ErrWideQueryResultsExhausted is used to short circuit additional document
+// entries being added if these wide results will no longer accept documents,
+// e.g. if the results are closed, or if no further documents will pass the
+// shard filter.
+var ErrWideQueryResultsExhausted = errors.New("no more values to add to wide query results")
+
+type shardFilterFn func(ident.ID) (uint32, bool)
+
+type wideResults struct {
+	sync.RWMutex
+	size           int
+	totalDocsCount int
+
+	nsID   ident.ID
+	idPool ident.Pool
+
+	closed      bool
+	idsOverflow []ident.ID
+	batch       *ident.IDBatch
+	batchCh     chan<- *ident.IDBatch
+	batchSize   int
+
+	shardFilter shardFilterFn
+	shards      []uint32
+	shardIdx    int
+	// NB: pastLastShard will mark this reader as exhausted after a
+	// document is discovered whose shard exceeds the last shard this results
+	// is responsible for, using the fact that incoming documents are sorted by
+	// shard then by ID.
+	pastLastShard bool
+}
+
+// NewWideQueryResults returns a new wide query results object.
+// NB: Reader must read results from `batchCh` in a goroutine, and call
+// batch.Done() after the result is used, and the writer must close the
+// channel after no more Documents are available.
+func NewWideQueryResults(
+	namespaceID ident.ID,
+	idPool ident.Pool,
+	shardFilter shardFilterFn,
+	collector chan *ident.IDBatch,
+	opts WideQueryOptions,
+) BaseResults {
+	batchSize := opts.BatchSize
+	results := &wideResults{
+		nsID:        namespaceID,
+		idPool:      idPool,
+		batchSize:   batchSize,
+		idsOverflow: make([]ident.ID, 0, batchSize),
+		batch: &ident.IDBatch{
+			IDs: make([]ident.ID, 0, batchSize),
+		},
+		batchCh: collector,
+		shards:  opts.ShardsQueried,
+	}
+
+	if len(opts.ShardsQueried) > 0 {
+		// Only apply filter if there are shards to filter against.
+		results.shardFilter = shardFilter
+	}
+
+	return results
+}
+
+func (r *wideResults) EnforceLimits() bool { return false }
+
+func (r *wideResults) AddDocuments(batch []doc.Document) (int, int, error) {
+	var size, totalDocsCount int
+	r.RLock()
+	size, totalDocsCount = r.size, r.totalDocsCount
+	r.RUnlock()
+
+	if r.closed || r.pastLastShard {
+		return size, totalDocsCount, ErrWideQueryResultsExhausted
+	}
+
+	r.Lock()
+	err := r.addDocumentsBatchWithLock(batch)
+	size, totalDocsCount = r.size, r.totalDocsCount
+	r.Unlock()
+
+	if err != nil && err != ErrWideQueryResultsExhausted {
+		// NB: if exhausted, drain the current batch and overflows.
+		return size, totalDocsCount, err
+	}
+
+	release := len(r.batch.IDs) == r.batchSize
```

Comment:
Does this need to be checked under RLock?

Reply:
Yeah will shuffle some things around, think i can make it a little clearer too